### PR TITLE
Added validations to the Results and SolveTime classes.

### PR DESCRIPTION
**Pull request by jfly**
When working on #977, I ended up doing this, and realized I could merge this up as is right now. I think smaller PRs are always a better thing, so here's a PR that gets us closer to porting the `check_results` script from PHP.

There were a couple of tricky things in here. One in particular is the `ResultMethods` module I created so that Result and LightResult can share the same functionality.

This is basically a port of https://github.com/thewca/worldcubeassociation.org/blob/master/webroot/results/includes/_check.php.

### CODE REVIEW

File: WcaOnRails/.rubocop.yml [hunk 1]
```diff
@@ -54,3 +54,9 @@ Style/SignalException:
 Style/ClassVars:
   Exclude:
     - app/models/country.rb
+
+Style/IfInsideElse:
+  Enabled: false
+
+Lint/NonLocalExitFromIterator:
+  Enabled: false
```
Comment (jfly) [inline WcaOnRails/.rubocop.yml]: Disabled because I can't think of a better way of writing: https://github.com/jfly/worldcubeassociation.org/blob/5868aef053809fdcf9d01776e7208b8e3d988cd0/WcaOnRails/app/models/result.rb#L40.

Comment (jfly) [inline WcaOnRails/.rubocop.yml]: Ahahaha, wow. Clever, but is it better? I'm going to sit on this for a day.

File: WcaOnRails/app/models/result.rb [hunk 2]
```diff
@@ -1,15 +1,86 @@
...
+
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+
+    if solve_times.all?(&:skipped?)
+      errors.add(:base, "Cannot skip all solves.")
+      return
```
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: Yeah, a number of our checks require knowing the format of the round and if the round is combined or not.

That refactoring looks nice. Will do.

Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Let me phrase that question clearer, less Seinfeldtesque way:

Normal validation functions collects all errors it can find. This one stops looking after finding the first error. What's the purpose of this unusual behavior?

File: WcaOnRails/app/models/result.rb [hunk 3]
```diff
@@ -1,15 +1,82 @@
...
+      sum_moves = 0
+      counting_solve_times.each do |solve_time|
+        sum_moves += solve_time.move_count
+      end
+      correct_average_wca_value = 100 * sum_moves / counting_solve_times.length
+    else
+      sum_centis = 0
+      counting_solve_times.each do |solve_time|
+        sum_centis += solve_time.time_centiseconds
+      end
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Rails also provides a nice `sum` method:

    sum_centis = counting_solve_times.sum(&: time_centiseconds)

File: WcaOnRails/.rubocop.yml (hunk 1 above)
Comment (larspetrus) [inline WcaOnRails/.rubocop.yml]: Yeah, it's fairly cryptic, and there has to be a better way, but to me it's an improvement.

I'll personally put up with a lot to get away from trailing pointer iterating...

File: WcaOnRails/app/models/result.rb (hunk 2 above)
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: Some issues arise if you try to keep checking for errors:

- We need a valid `roundId` and `formatId` to know how many solves we expect the Result to have. Without those, we could check the `best`, but we cannot check the `average`.
- If a round only had 3 solves, and there are 4 solves entered, I don't know what it would mean for us to check the `average`.
- If a round had 5 solves, but some of them were skipped, I don't know what it would mean for us to check the `average`.
- We need to know the `eventId` to compute `average`, because 333fm's average field is so funky.

Basically, there are some preconditions we need to satisfy before it makes sense to talk about the result's "average". Maybe there's a better way of thinking of this, let me know what you think.

3We could rewrite this to always check `best`.

Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Aha. That makes sense!

I don't think it applies to all the returns, like the `nil` checks in line 19-23. Some of the others don't look forced either, but I don't know enough to say for sure.

The way to get around having returns all over would be making them an array of Lambdas that you execute until you've found an error. Or you can set booleans instead of returning. `average_is_computable = false`, and then wrap the rest of the code in `if average_is_computable`, etc. Both strike me as a cure worse than the disease.

Anyway. I think it would help a lot with readability to have a short comment explaining each exit. If someone needs to add a new validation, it would be pretty hard to understand where in the hierarchy to insert it otherwise.

File: WcaOnRails/app/models/result.rb [hunk 4]
```diff
@@ -1,15 +1,82 @@
...
   belongs_to :person, -> { current }, primary_key: :wca_id, foreign_key: :personId
   belongs_to :round, foreign_key: :roundId
   belongs_to :event, foreign_key: :eventId
+  belongs_to :format, foreign_key: :formatId
 
   scope :podium, -> { joins(:round).merge(Round.final_rounds).where(pos: [1..3]).where("best > 0") }
   scope :winners, -> { joins(:round, :event).merge(Round.final_rounds).where("pos = 1 and best > 0").order("Events.rank") }
 
+  validate :best_and_average
+  def best_and_average
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: The function validates a lot more than the name suggests.

You don't have to make it a named function. `validate do... end` works.

File: WcaOnRails/app/models/result.rb [hunk 5]
```diff
@@ -1,15 +1,82 @@
...
+      counting_solve_times.each do |solve_time|
+        sum_centis += solve_time.time_centiseconds
+      end
+      correct_average_wca_value = sum_centis / counting_solve_times.length
+    end
+
+    if correct_average_wca_value != average
+      errors.add(:average, "should be #{correct_average_wca_value}")
+      return
+    end
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: In the earlier comment I meant - but didn't say - that i like these patterns applied to the whole function

`return errors.add(...)`
`return errors.add(...) if condition`

File: WcaOnRails/app/models/result.rb (hunk 4 above)
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: WOW seriously?? I have been naming these things for no reason...

Comment (jfly) [inline WcaOnRails/app/models/result.rb]: Ok, this gets interesting. If I use the `validate do... end` pattern, I get a

```ruby
     LocalJumpError:
       unexpected return
```

Apparently you're only allowed to do a `return` in ruby if you're inside of a method, but with the `validate do... end` pattern, there is no enclosing method, just a block.

Comment (jonatanklosko) [inline WcaOnRails/app/models/result.rb]: Try to use a lambda instead.
```ruby
validate -> do
# Cool stuff using `return`
end
```

File: WcaOnRails/app/models/result.rb (hunk 5 above)
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Agreed

File: WcaOnRails/app/models/result.rb (hunk 2 above)
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: In light of this discussion, I ended up refactoring this a good bit by pulling things out into separate validations. I think this is slightly more readable. Let me know what you think!

Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Splitting it does make each function easier to digest.

Personally I'd have a main validation function that calls all the smaller ones, to kinda collect it all in one place. But that may not be The Rails Way.

File: WcaOnRails/app/models/result.rb [hunk 6]
```diff
@@ -1,15 +1,85 @@
...
+
+    unskipped_count = solve_times.length - solve_times.count(&:skipped?)
+    if round.combined?
+      if unskipped_count > format.expected_solve_count
+        return errors.add(:base, "Expected at most #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
+      end
+    else
+      if unskipped_count != format.expected_solve_count
+        return errors.add(:base, "Expected #{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}, but found #{unskipped_count}.")
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: I realized why these `pluralize` calls look so weird to me. Apparently, there are two `pluralize` functions in Rails!  The other one is better: With http://apidock.com/rails/ActionView/Helpers/TextHelper/pluralize, this

`#{format.expected_solve_count} #{'solve'.pluralize(format.expected_solve_count)}`

becomes this

`#{pluralize(format.expected_solve_count, 'solve')}`

I put this in my models when I need helper functions:

    def hlp
      ActionController::Base.helpers
    end

File: WcaOnRails/.rubocop.yml (hunk 1 above)
Comment (larspetrus) [inline WcaOnRails/.rubocop.yml]: Another option:

    last_unskipped_index = solve_times.rindex{|st|  not st.skipped? } || 0
    first_skipped_index = solve_times..index{|st| st.skipped? } || 999
    errors.add(:base, "Skipped solves must all come at the end.") if last_unskipped_index > first_skipped_index

Comment (jfly) [inline WcaOnRails/.rubocop.yml]: I like this! I decided to use `Float::INFINITY` instead of `999`, however.

Comment (vincentwoo) [inline WcaOnRails/.rubocop.yml]: Please allow me to contribute what I consider to be an idiomatic ruby solution:

```ruby
skipped_index = solve_times.index(&:skipped?) || solve_times.length
unless solve_times[skipped_index..solve_times.length].all?(&:skipped?)
  errors.add(:base, "Skipped solves must all come at the end.")
end
```

Comment (jfly) [inline WcaOnRails/.rubocop.yml]: WOW. Ok, I *really* like @vincentwoo's solution. I think it follows very clearly from the problem definition.

Comment (vincentwoo) [inline WcaOnRails/.rubocop.yml]: holy shit forget that, `drop_while` is the shit. adapting that other guy's solution:

```ruby
solve_times.drop_while { |st| !st.skipped? }.all?(&:skipped)
```

Comment (jfly) [inline WcaOnRails/.rubocop.yml]: Good call! Thank you @vincentwoo and @jonatanklosko I couldn't resist adding a `unskipped` method, so the solution becomes:

```ruby
solve_times.drop_while(&:unskipped?).all?(&:skipped?)
```

File: WcaOnRails/.rubocop.yml [hunk 7]
```diff
@@ -54,3 +54,9 @@ Style/SignalException:
 Style/ClassVars:
   Exclude:
     - app/models/country.rb
+
+Style/IfInsideElse:
+  Enabled: false
+
+Lint/NonLocalExitFromIterator:
```
Comment (larspetrus) [inline WcaOnRails/.rubocop.yml]: So if these rules don't need to be broken anymore, they should probably be reinstated.

File: WcaOnRails/app/models/result.rb [hunk 8]
```diff
@@ -1,16 +1,105 @@
...
+    end
+  end
+
+  validate :validate_number_of_solves
+  def validate_number_of_solves
+    return errors.add(:competitionId, "invalid") unless competition
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: I still don't see why these checks should be mutually exclusive.

File: WcaOnRails/app/models/result.rb [hunk 9]
```diff
@@ -1,16 +1,105 @@
...
+  def validate_each_solve
+    solve_times.each_with_index do |solve_time, i|
+      unless solve_time.valid?
+        errors.add(:"value#{i + 1}", solve_time.errors.messages[:base].join(" "))
+      end
+    end
+  end
+
+  validate :validate_number_of_solves
+  def validate_number_of_solves
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: The name describes only one of the 6 checks here.

File: WcaOnRails/app/models/result.rb [hunk 10]
```diff
@@ -1,16 +1,105 @@
...
+    return errors.add(:countryId, "invalid") unless country
+    return errors.add(:roundId, "invalid") unless round
+    return errors.add(:eventId, "invalid") unless event
+    return errors.add(:formatId, "invalid") unless format
+    return errors.add(:base, invalid_number_of_solves_reason) if invalid_number_of_solves_reason
+  end
+
+  def invalid_number_of_solves_reason
+    return "Invalid format" unless format
+    return "Invalid round" unless round
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: These two are duplicates.

File: WcaOnRails/app/models/result.rb [hunk 11]
```diff
@@ -1,16 +1,105 @@
...
+    return "Cannot skip all solves." if solve_times.all?(&:skipped?)
+
+    unless solve_times.drop_while(&:unskipped?).all?(&:skipped?)
+      return "Skipped solves must all come at the end."
+    end
+
+    unskipped_count = solve_times.length - solve_times.count(&:skipped?)
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Use `unskipped?` now that we have it.

File: WcaOnRails/app/models/result.rb [hunk 12]
```diff
@@ -1,16 +1,105 @@
...
+    correct_best = compute_correct_best
+    errors.add(:best, "should be #{correct_best}") if correct_best != best
+  end
+
+  def compute_correct_best
+    sorted_solves.first&.wca_value
+  end
+
+  def compute_correct_average
+    fail average_is_not_computable_reason if average_is_not_computable_reason
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: Raising an exception in the validation code seems wrong?

File: WcaOnRails/app/models/result.rb [hunk 13]
```diff
@@ -1,16 +1,105 @@
...
+        return "Expected #{hlp.pluralize(format.expected_solve_count, 'solve')}, but found #{unskipped_count}."
+      end
+    end
+  end
+
+  def average_is_not_computable_reason
+    # To compute the average, we need to have a valid number of solves,
+    # and we need to know what event we are dealing with (because
+    # 333fm is computed differently than other events).
+    !event ? "Event needed to compute average" : invalid_number_of_solves_reason
```
Comment (larspetrus) [inline WcaOnRails/app/models/result.rb]: 1. `!event ? a : b` should be `event ? b : a`

2. I'm really struggling to follow this code. These seem to be chained `*_reason` methods that return strings but are treated as booleans.

File: WcaOnRails/.rubocop.yml (hunk 7 above)
Comment (jfly) [inline WcaOnRails/.rubocop.yml]: We're still breaking the `Style/IfInsideElse` rule [here](https://github.com/jfly/worldcubeassociation.org/blob/56182b003d0bd370e19db334d4513d4406fe58df/WcaOnRails/app/models/result.rb#L50-L54) and [here](https://github.com/jfly/worldcubeassociation.org/blob/56182b003d0bd370e19db334d4513d4406fe58df/WcaOnRails/app/models/result.rb#L87-L95), but I like breaking it to achieve parallelism.

I have removed the `Lint/NonLocalExitFromIterator`.

File: WcaOnRails/app/models/result.rb (hunk 8 above)
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: Ah, my bad. I've pulled out the `countryId`, `competitionId`, and `eventId` checks, because those are irrelevant for validating the number of solves.

However, we *do* need to know the `roundId` and the `formatId` in order to validate the number of solves. I've made a change that I hope makes this a little clearer clearer.

File: WcaOnRails/app/models/result.rb (hunk 12 above)
Comment (jfly) [inline WcaOnRails/app/models/result.rb]: This is actually an assertion. Someone calling `compute_correct_average` is expected to first check if the average is actually computable.

I'm definitely open to restructuring this. Are you free to chat today?

**Comment by larspetrus** [comment]
This is still not how I would write it in some ways, but the code works and let's not have our different preferences block this any further.

LGTM!

**Comment by jfly** [comment]
Ok, thanks for the reviews, @larspetrus! This is definitely better than the first iteration =)
